crates/core/src/instruction.rs

```diff
@@ -61,6 +61,115 @@ pub struct InstructionMetadata {
+    ///
+    /// This method filters the transaction's log messages to return only those
+    /// that correspond to the current instruction, based on its stack height and
+    /// absolute path within the instruction stack.
+    ///
+    /// Returns `Vec<Vec<u8>>` containing the `data` bytes (base64 encoded) from log messages.
+    pub fn extract_logs(&self) -> Vec<Vec<u8>> {
+        let logs = match &self.transaction_metadata.meta.log_messages {
+            Some(logs) => logs,
+            None => return Vec::new(),
+        };
+
+        let mut extracted_logs = Vec::new();
+        let mut current_stack_height = 0usize;
+        let mut last_stack_height = 0usize;
+
+        let mut position_at_level: std::collections::HashMap<usize, u8> =
+            std::collections::HashMap::new();
+
+        for log in logs {
+            let parsed_log = self.parse_log(log);
+
+            match parsed_log {
+                LogType::Start(stack_height) => {
+                    current_stack_height = stack_height;
+
+                    // If we're at a new level higher than before, initialize at 0
+                    // If we're at same or lower level, increment
+                    let current_pos = if stack_height > last_stack_height {
+                        // Going higher - start at 0
+                        0
+                    } else {
+                        // Same level or going back down - increment from last position
+                        position_at_level
+                            .get(&stack_height)
+                            .map(|&pos| pos + 1)
+                            .unwrap_or(0)
+                    };
+
+                    position_at_level.insert(stack_height, current_pos);
+                    last_stack_height = stack_height;
+                }
+                LogType::Finish => {
+                    // When finishing, we go back to parent level
+                    current_stack_height = current_stack_height.saturating_sub(1);
+                }
+                // No changes for Data/CU
+                _ => {}
+            }
+
+            // Build current path from position tracking
+            let current_path: Vec<u8> = (1..=current_stack_height)
+                .map(|level| position_at_level.get(&level).copied().unwrap_or(0))
+                .collect();
+
+            // Extract if we're on target path and it's a data log
+            if current_path == self.absolute_path && matches!(parsed_log, LogType::Data) {
+                if let Some(data) = log.split_whitespace().last() {
+                    if let Ok(buf) =
+                        base64::Engine::decode(&base64::engine::general_purpose::STANDARD, data)
+                    {
+                        extracted_logs.push(buf);
+                    }
+                }
+            }
+        }
+
+        extracted_logs
+    }
+
+    /// Parses a log line to determine its type
+    fn parse_log(&self, log: &str) -> LogType {
```

Comment:
Hmm I wonder, if there is not something already in anchor or some solana library that can do this decoding. This seems fickle.